kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java

```diff
@@ -58,16 +56,50 @@ public void initChannel(SocketChannel ch) {
 
         ChannelPipeline pipeline = ch.pipeline();
 
-        sslContext.ifPresent(s -> {
+        int targetPort = ch.localAddress().getPort();
+        if (requireTls) {
             LOGGER.debug("Adding SSL/SNI handler");
-            pipeline.addLast(new SniHandler(new DomainWildcardMappingBuilder<>(s).build()));
-        });
+            pipeline.addLast(new SniHandler(hostname -> {
+                var virtualCluster = virtualClusterResolver.resolve(hostname, targetPort);
+                var sslContext = virtualCluster.buildSslContext();
+                if (sslContext.isEmpty()) {
+                    throw new IllegalStateException("Virtual cluster %s does not provide SSL context".formatted(virtualCluster));
+                }
+                ch.attr(VIRTUAL_CLUSTER).set(virtualCluster);
```

Comment:
Maybe we could avoid passing the state via channel attribute and the additional event handler by adding the Handlers as part of this mapping function, and override `SniHandler#onLookupComplete` to call `ctx.fireChannelActive()`.

Like:
```java
pipeline.addLast(new SniHandler(hostname -> {
    var virtualCluster = virtualClusterResolver.resolve(hostname, targetPort);
    ... snip ...
    addHandlers(ch, virtualCluster);
    return sslContext.get();
}) {
    @Override
    protected void onLookupComplete(ChannelHandlerContext ctx, Future<SslContext> future) throws Exception {
        super.onLookupComplete(ctx, future);
        ctx.fireChannelActive();
    }
});
```

Reply:
@robobario yeah that work's. Much less hocus pocus! Thank you!